# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/gtk/WakeLockListener.cpp
# Commit: 2a39a8e425b0
# Full Hash: 2a39a8e425b0824c670ba03452c437c3748c990b
# Author: Greg V <greg@unrelenting.technology>
# Date: 2020-01-06 21:54:03
# Regressor Bug: 1587360
# File Overlap Count: 1
# Description:
#   Bug 1587360 - Add support for idle-inhibit Wayland protocol r=stransky
#   
#   The idle-inhibit protocol allows idle wake locks on Wayland
#   without any of the supported D-Bus interfaces running.
#   
# ==============================================================================

diff -r 47592189b305 -r 2a39a8e425b0 widget/gtk/WakeLockListener.cpp
--- a/widget/gtk/WakeLockListener.cpp	Mon Jan 06 17:02:14 2020 +0000
+++ b/widget/gtk/WakeLockListener.cpp	Mon Jan 06 16:11:19 2020 +0000
@@ -16,6 +16,12 @@
 #    include "prlink.h"
 #  endif
 
+#  if defined(MOZ_WAYLAND)
+#    include "mozilla/widget/nsWaylandDisplay.h"
+#    include "nsWindow.h"
+#    include "mozilla/dom/power/PowerManagerService.h"
+#  endif
+
 #  define FREEDESKTOP_SCREENSAVER_TARGET "org.freedesktop.ScreenSaver"
 #  define FREEDESKTOP_SCREENSAVER_OBJECT "/ScreenSaver"
 #  define FREEDESKTOP_SCREENSAVER_INTERFACE "org.freedesktop.ScreenSaver"
@@ -38,18 +44,26 @@
 #  if defined(MOZ_X11)
   XScreenSaver,
 #  endif
+#  if defined(MOZ_WAYLAND)
+  WaylandIdleInhibit,
+#  endif
   Unsupported,
 };
 
 class WakeLockTopic {
  public:
   WakeLockTopic(const nsAString& aTopic, DBusConnection* aConnection)
-      : mTopic(NS_ConvertUTF16toUTF8(aTopic)),
+      :
+#  if defined(MOZ_WAYLAND)
+        mWaylandInhibitor(nullptr),
+#  endif
+        mTopic(NS_ConvertUTF16toUTF8(aTopic)),
         mConnection(aConnection),
         mDesktopEnvironment(FreeDesktop),
         mInhibitRequest(0),
         mShouldInhibit(false),
-        mWaitingForReply(false) {}
+        mWaitingForReply(false) {
+  }
 
   nsresult InhibitScreensaver(void);
   nsresult UninhibitScreensaver(void);
@@ -67,6 +81,13 @@
   static bool InhibitXScreenSaver(bool inhibit);
 #  endif
 
+#  if defined(MOZ_WAYLAND)
+  zwp_idle_inhibitor_v1* mWaylandInhibitor;
+  static bool CheckWaylandIdleInhibitSupport();
+  bool InhibitWaylandIdle();
+  bool UninhibitWaylandIdle();
+#  endif
+
   static void ReceiveInhibitReply(DBusPendingCall* aPending, void* aUserData);
   void InhibitFailed();
   void InhibitSucceeded(uint32_t aInhibitRequest);
@@ -197,6 +218,49 @@
 
 #  endif
 
+#  if defined(MOZ_WAYLAND)
+
+/* static */
+bool WakeLockTopic::CheckWaylandIdleInhibitSupport() {
+  GdkDisplay* gDisplay = gdk_display_get_default();
+  if (GDK_IS_X11_DISPLAY(gDisplay)) return false;
+
+  widget::nsWaylandDisplay* waylandDisplay =
+      widget::WaylandDisplayGet(gDisplay);
+  if (!waylandDisplay->GetIdleInhibitManager()) return false;
+
+  return true;
+}
+
+bool WakeLockTopic::InhibitWaylandIdle() {
+  GdkDisplay* gDisplay = gdk_display_get_default();
+
+  widget::nsWaylandDisplay* waylandDisplay =
+      widget::WaylandDisplayGet(gDisplay);
+
+  nsWindow* focusedWindow = nsWindow::GetFocusedWindow();
+  if (!focusedWindow) return false;
+
+  UninhibitWaylandIdle();
+
+  mWaylandInhibitor = zwp_idle_inhibit_manager_v1_create_inhibitor(
+      waylandDisplay->GetIdleInhibitManager(),
+      focusedWindow->GetWaylandSurface());
+
+  return true;
+}
+
+bool WakeLockTopic::UninhibitWaylandIdle() {
+  if (mWaylandInhibitor == nullptr) return false;
+
+  zwp_idle_inhibitor_v1_destroy(mWaylandInhibitor);
+  mWaylandInhibitor = nullptr;
+
+  return true;
+}
+
+#  endif
+
 bool WakeLockTopic::SendInhibit() {
   bool sendOk = false;
 
@@ -211,6 +275,10 @@
     case XScreenSaver:
       return InhibitXScreenSaver(true);
 #  endif
+#  if defined(MOZ_WAYLAND)
+    case WaylandIdleInhibit:
+      return InhibitWaylandIdle();
+#  endif
     case Unsupported:
       return false;
   }
@@ -239,6 +307,11 @@
     return InhibitXScreenSaver(false);
   }
 #  endif
+#  if defined(MOZ_WAYLAND)
+  else if (mDesktopEnvironment == WaylandIdleInhibit) {
+    return UninhibitWaylandIdle();
+  }
+#  endif
 
   if (!message) {
     return false;
@@ -300,6 +373,10 @@
   } else if (mDesktopEnvironment == GNOME && CheckXScreenSaverSupport()) {
     mDesktopEnvironment = XScreenSaver;
 #  endif
+#  if defined(MOZ_WAYLAND)
+  } else if (mDesktopEnvironment == GNOME && CheckWaylandIdleInhibitSupport()) {
+    mDesktopEnvironment = WaylandIdleInhibit;
+#  endif
   } else {
     mDesktopEnvironment = Unsupported;
     mShouldInhibit = false;